cmd/test-generation/generate-test.go

```diff
@@ -0,0 +1,342 @@
+	uniqueMap := make(map[string]bool)
+	uniques := make([]string, 0, len(expectations))
+
+	for id, replicatees := range expectations {
+		if _, ok := uniqueMap[id]; !ok && len(replicatees) > 0 {
+			uniques = append(uniques, id)
+			uniqueMap[id] = true
+		}
+		for _, replicateeId := range replicatees {
+			if _, ok := uniqueMap[replicateeId]; !ok {
+				uniques = append(uniques, replicateeId)
+				uniqueMap[replicateeId] = true
+			}
+		}
+	}
+	return uniques
+}
+
+type runtimeArgs struct {
+	ssbServer        string
+	fixturesRoot     string
+	expectationsPath string
+	batchSize        int
+	focusedPuppets   int
+	maxHops          int
+}
+
+// uses:
+// * expectations.json
+// * root folder containing cmd&log-splicer processed fixtures
+var currentlyExecuting map[string]bool
+var idsToNames map[string]string
+var focusGroup []string
+var isBlocking map[string]map[string]bool
+var args runtimeArgs
+
+func main() {
+	currentlyExecuting = make(map[string]bool)
+	flag.StringVar(&args.fixturesRoot, "fixtures", "./fixtures-output", "root folder containing spliced out ssb-fixtures")
+	flag.StringVar(&args.expectationsPath, "expectations", "./expectations.json", "path to expectations.json")
+	flag.StringVar(&args.ssbServer, "sbot", "ssb-server", "the ssb server to start puppets with")
+	flag.IntVar(&args.maxHops, "hops", 2, "the max hops count to use")
+	flag.IntVar(&args.focusedPuppets, "focused", 2, "number of puppets to use for focus group (i.e. # of puppets that verify they are replicating others)")
+	flag.Parse()
+
+	expectations, err := readExpectations(args.expectationsPath)
+	check(err)
+
+	// map of id -> [list of followed ids]
+	var followMap map[string][]string
+	followMap, isBlocking, err = getFollowMap(path.Join(args.fixturesRoot, "follow-graph.json"))
+	check(err)
+
+	idsToNames, err = getIdentities(args.fixturesRoot)
+	check(err)
+
+	puppetNames := make([]string, 0, len(idsToNames))
+	namesToIDs := make(map[string]string)
+	for id, secretFolder := range idsToNames {
+		namesToIDs[secretFolder] = id
+		puppetNames = append(puppetNames, idsToNames[id])
+	}
+	sort.Strings(puppetNames)
+
+	getNames := func(src []string) []string {
+		extractedNames := make([]string, 0, len(src))
+		for _, id := range src {
+			extractedNames = append(extractedNames, idsToNames[id])
+		}
+		return extractedNames
+	}
+
+	// Note: do not remove, it's a useful help function: )
+	getIds := func(src []string) []string {
+		extractedIds := make([]string, 0, len(src))
+		for _, name := range src {
+			extractedIds = append(extractedIds, namesToIDs[name])
+		}
+		return extractedIds
+	}
+
+	// the cohort of peers we care about; the ones who will be issuing `has` stmts, the ones whose data we will inspect
+	focusGroup = make([]string, args.focusedPuppets)
+	for i := 0; i < args.focusedPuppets; i++ {
+		focusGroup[i] = fmt.Sprintf("puppet-%05d", i)
+	}
+	// deterministically shuffle the focus group
+	// TODO: accept --seed flag to change shuffling
```

Comment:
i guess you've seen this already but just in case. Using [rand.New](https://pkg.go.dev/math/rand#New)` you can feed in a seed and call `Shuffle()` on the new randonness thing to achive this.